packages/tasit-sdk/src/Decentraland.test.js

```diff
@@ -136,111 +136,176 @@ describe("Decentraland", () => {
+            );
+            await expectExactTokenBalances(
+              mana,
+              [ephemeralAddress],
+              [manaAmountForShopping]
+            );
+
+            mana.setWallet(ephemeralWallet);
+            const approvalAction = mana.approve(
+              MARKETPLACE_ADDRESS,
+              manaAmountForShopping
+            );
+
+            const errorListener = error => {
+              marketplace.off("error");
+              done(error);
+            };
+
+            const successfulListener = async message => {
+              const allowance = await mana.allowance(
+                ephemeralAddress,
+                MARKETPLACE_ADDRESS
+              );
+              expect(`${allowance}`).to.equal(`${manaAmountForShopping}`);
+              done();
+            };
+
+            mana.once("Approval", successfulListener);
+            mana.on("error", errorListener);
+
+            approvalAction.send();
+          })();
         });
 
-        it("should buy an estate", async () => {
-          const {
-            assetId,
-            nftAddress,
-            seller,
-            priceInWei,
-            expiresAt,
-          } = estateForSale;
-
-          await checkAsset(estate, mana, estateForSale, ephemeralAddress);
-
-          await expectExactTokenBalances(estate, [ephemeralAddress], [0]);
-
-          const fingerprint = await estate.getFingerprint(`${assetId}`);
-
-          marketplace.setWallet(ephemeralWallet);
-          const executeOrderAction = marketplace.safeExecuteOrder(
-            nftAddress,
-            `${assetId}`,
-            `${priceInWei}`,
-            `${fingerprint}`
-          );
-          await executeOrderAction.send();
-          await executeOrderAction.waitForOneConfirmation();
-
-          await expectExactTokenBalances(estate, [ephemeralAddress], [1]);
+        it("should buy an estate", done => {
+          (async () => {
+            const {
+              assetId,
+              nftAddress,
+              seller,
+              priceInWei,
+              expiresAt,
+            } = estateForSale;
+
+            await checkAsset(estate, mana, estateForSale, ephemeralAddress);
+
+            await expectExactTokenBalances(estate, [ephemeralAddress], [0]);
+
+            const fingerprint = await estate.getFingerprint(`${assetId}`);
+
+            marketplace.setWallet(ephemeralWallet);
+            const executeOrderAction = marketplace.safeExecuteOrder(
+              nftAddress,
+              `${assetId}`,
+              `${priceInWei}`,
+              `${fingerprint}`
+            );
+
+            const orderSuccessfulListener = async message => {
+              const { data } = message;
+              const { args } = data;
+              const { buyer } = args;
+
+              expect(buyer).to.equal(ephemeralAddress);
+              await expectExactTokenBalances(estate, [ephemeralAddress], [1]);
+
+              done();
```

Comment:
If this were working properly and it wasn't commented out where this was added, we wouldn't actually want the test to race for two `done`s, right? Instead we would split this out into two tests or pick just one style? Or not call done until both were true?

Reply:
Are you saying about both `orderSuccessfulListener` and `confirmationListener` on the same test case? If so, we are using only the `confirmationListener`. The intention is to use the `OrderSucessful` event to trigger listener but it isn't working right now (https://github.com/tasitlabs/TasitSDK/issues/367). Because of that, I've added the `confirmationListener` for now.

Reply:
Yeah, I'm saying the way it's written right now it's not clear if you'd use both listeners in the same test case, because we kind of are now even though one is commented out.

I think my main preference would be a unit test somewhere that confirms that either method would work in two separate tests. And then in our functional tests throughout the rest of the code, we just use events like you're suggesting.

Maybe include a comment in the code to signal this intent to only use events here in the future?

Reply:
I've splitted the test case and now we have one using contract events and another using action events.

Reply:
Great, thanks!! Resolving this